### PR TITLE
Fix ANCM environment variables bugs

### LINKED ISSUE #4566: Duplicate ASPNETCORE_Environment present for InProcess hosting causes ASPNET CORE 2.2 fails to load appsettings
After 2 days trying to figure out, i think i discovered a bug:

1) I already had set manually the ASPNETCORE_ENVIRONMENT on Windows Server 2016 environment variables.
2) After 2.2 release, i decided to update the publish profile .pubxml to include the desired environment, for instance: 

`<EnvironmentName>Development</EnvironmentName>`

**This generates the web.config:**
```
<aspNetCore processPath="dotnet" arguments=".\TJPA.CentralBoletos.API.dll" stdoutLogEnabled="true" stdoutLogFile=".\logs\stdout" hostingModel="inprocess">
   <environmentVariables>
           <environmentVariable name="ASPNETCORE_ENVIRONMENT" value="Development" />
   </environmentVariables>
</aspNetCore>
```
 

3) When i published the application, the application wont startup, specifically when reading the 
connection string for the HangfireDB schema i was using. 

**In Web browser says:**

> An error occurred while starting the application.
> .NET Core 4.6.27110.04 X64 v4.0.0.0    |   Microsoft.AspNetCore.Hosting version 2.2.0-rtm-35687    |    Microsoft Windows 10.0.14393    |   Need help?

**In std logs:**

> Application startup exception: System.ArgumentNullException: Value cannot be null.
> Parameter name: nameOrConnectionString
>    at Hangfire.SqlServerStorageExtensions.UseSqlServerStorage(IGlobalConfiguration configuration, String nameOrConnectionString, SqlServerStorageOptions options)
>    at TJPA.CentralBoletos.API.Startup.<>c__DisplayClass5_0.<ConfigureServices>b__2(IGlobalConfiguration config) in C:\GIT\central-boletos\api-central-boletos\TJPA.CentralBoletos.API\Startup.cs:line 52
>    at Hangfire.HangfireServiceCollectionExtensions.<>c__DisplayClass0_0.<AddHangfire>b__9(IServiceProvider serviceProvider)
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)
>    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
>    at Hangfire.HangfireApplicationBuilderExtensions.ThrowIfNotConfigured(IApplicationBuilder app)
>    at Hangfire.HangfireApplicationBuilderExtensions.UseHangfireServer(IApplicationBuilder app, BackgroundJobServerOptions options, IEnumerable`1 additionalProcesses, JobStorage storage)
>    at TJPA.CentralBoletos.API.Startup.ConfigurarHangfireServer(IApplicationBuilder app, IServiceProvider serviceProvider) in C:\GIT\central-boletos\api-central-boletos\TJPA.CentralBoletos.API\Startup.cs:line 149
>    at TJPA.CentralBoletos.API.Startup.Configure(IApplicationBuilder app, IServiceProvider serviceProvider) in C:\GIT\central-boletos\api-central-boletos\TJPA.CentralBoletos.API\Startup.cs:line 79
> --- End of stack trace from previous location where exception was thrown ---
>    at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.Configure(IApplicationBuilder app)
>    at Microsoft.AspNetCore.Server.IIS.Core.IISServerSetupFilter.<>c__DisplayClass2_0.<Configure>b__0(IApplicationBuilder app)
>    at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
> **Hosting environment: Development;Development**
> Content root path: C:\inetpub\wwwroot\central-boletos
> Application started. Press Ctrl+C to shut down.
> 

**The ancm log:**

> [aspnetcorev2.dll] Initializing logs for 'C:\Program Files\IIS\Asp.Net Core Module\V2\aspnetcorev2.dll'. Process Id: 10784.. File Version: 12.2.18316.0. Description: IIS ASP.NET Core Module V2. Commit: ce8cf65589734f82b0536c543aba5bd60d0a5a98.
> [aspnetcorev2.dll] Resolving hostfxr parameters for application: 'dotnet' arguments: 'TJPA.CentralBoletos.API.dll' path: 'C:\inetpub\wwwroot\central-boletos\'
> [aspnetcorev2.dll] Known dotnet.exe location: ''
> [aspnetcorev2.dll] Process path 'dotnet.exe' is dotnet, treating application as portable
> [aspnetcorev2.dll] Resolving absolute path to dotnet.exe from 'dotnet.exe'
> [aspnetcorev2.dll] Invoking where.exe to find dotnet.exe
> [aspnetcorev2.dll] where.exe invocation returned: 'C:\Program Files\dotnet\dotnet.exe
> C:\Program Files (x86)\dotnet\dotnet.exe
> '
> [aspnetcorev2.dll] Current process bitness type detected as isX64=1
> [aspnetcorev2.dll] Processing entry 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2.dll] Binary type 6
> [aspnetcorev2.dll] Found dotnet.exe via where.exe invocation at 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2.dll] Resolving absolute path to hostfxr.dll from 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2.dll] hostfxr.dll located at 'C:\Program Files\dotnet\host\fxr\2.2.0\hostfxr.dll'
> [aspnetcorev2.dll] Converted argument 'TJPA.CentralBoletos.API.dll' to 'C:\inetpub\wwwroot\central-boletos\TJPA.CentralBoletos.API.dll'
> [aspnetcorev2.dll] Parsed hostfxr options: dotnet location: 'C:\Program Files\dotnet\dotnet.exe' hostfxr path: 'C:\Program Files\dotnet\host\fxr\2.2.0\hostfxr.dll' arguments:
> [aspnetcorev2.dll] Argument[0] = 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2.dll] Argument[1] = 'C:\inetpub\wwwroot\central-boletos\TJPA.CentralBoletos.API.dll'
> [aspnetcorev2.dll] c:\b\w\e37dd45d8cd1eaf4\src\iisintegration\src\aspnetcoremodulev2\commonlib\fileoutputmanager.cpp:142 Operation failed with LastError: 32 HR: 0x80070020
> [aspnetcorev2.dll] Loading request handler:  'C:\inetpub\wwwroot\central-boletos\runtimes\win-x64\nativeassets\netcoreapp2.2\aspnetcorev2_inprocess.dll'
> [aspnetcorev2.dll] Creating handler application
> [aspnetcorev2_inprocess.dll] Initializing logs for 'C:\inetpub\wwwroot\central-boletos\runtimes\win-x64\nativeassets\netcoreapp2.2\aspnetcorev2_inprocess.dll'. Process Id: 10784.. File Version: 12.2.18316.0. Description: IIS ASP.NET Core Module V2 Request Handler. Commit: ce8cf65589734f82b0536c543aba5bd60d0a5a98.
> [aspnetcorev2_inprocess.dll] Waiting for initialization
> [aspnetcorev2_inprocess.dll] Starting in-process worker thread
> [aspnetcorev2_inprocess.dll] Resolving hostfxr parameters for application: 'dotnet' arguments: 'TJPA.CentralBoletos.API.dll' path: 'C:\inetpub\wwwroot\central-boletos\'
> [aspnetcorev2_inprocess.dll] Known dotnet.exe location: 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2_inprocess.dll] Process path 'dotnet.exe' is dotnet, treating application as portable
> [aspnetcorev2_inprocess.dll] Resolving absolute path to hostfxr.dll from 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2_inprocess.dll] hostfxr.dll located at 'C:\Program Files\dotnet\host\fxr\2.2.0\hostfxr.dll'
> [aspnetcorev2_inprocess.dll] Converted argument 'TJPA.CentralBoletos.API.dll' to 'C:\inetpub\wwwroot\central-boletos\TJPA.CentralBoletos.API.dll'
> [aspnetcorev2_inprocess.dll] Parsed hostfxr options: dotnet location: 'C:\Program Files\dotnet\dotnet.exe' hostfxr path: 'C:\Program Files\dotnet\host\fxr\2.2.0\hostfxr.dll' arguments:
> [aspnetcorev2_inprocess.dll] Argument[0] = 'C:\Program Files\dotnet\dotnet.exe'
> [aspnetcorev2_inprocess.dll] Argument[1] = 'C:\inetpub\wwwroot\central-boletos\TJPA.CentralBoletos.API.dll'
> [aspnetcorev2_inprocess.dll] In-process callbacks set
> [aspnetcorev2_inprocess.dll] Event Log: 'Application 'C:\inetpub\wwwroot\central-boletos\' started the coreclr in-process successfully.' 
> End Event Log Message.
> [aspnetcorev2_inprocess.dll] Starting app_offline monitoring in application 'C:\inetpub\wwwroot\central-boletos\'
> [aspnetcorev2_inprocess.dll] Starting file watcher thread

When i removed the lines in web.config solved the problem:
```
   <environmentVariables>
           <environmentVariable name="ASPNETCORE_ENVIRONMENT" value="Development" />
   </environmentVariables>
```

### PR DESCRIPTION
**Pull request by pakrym**
New rules:
1. web.config overrides everything but ASPNETCORE_IIS_HTTPAUTH and ASPNETCORE_HOSTINGSTARTUPASSEMBLIES
2.  `ASPNETCORE_IIS_HTTPAUTH` is always overridden by the ANCM
3. For `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` web.config value overrides the global environment variable. In out of proc case `Microsoft.AspNetCore.Server.IISIntegration` is appended to the global environment variable if web.config doesn't have one.

Fixes: https://github.com/aspnet/AspNetCore/issues/4566

### CODE REVIEW

**Comment by shirhatti** [comment]
Just `ASPNETCORE_IIS_HTTPAUTH` or should we establish a well-known prefix? Will probably save us from a shim change in the future?

**Comment by pakrym** [comment]
I'm keeping `ASPNETCORE_IIS_HTTPAUTH` behaviour for backwards compatibility. I doubt we would need to add any other variable like this in the future.

File: .azure/pipelines/ci.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ jobs:
 - template: jobs/iisintegration-job.yml
   parameters:
     TestGroupName: IISExpress
-    skipArgs: "/p:SkipIISBackwardsCompatibilityTests=false /p:SkipIISTests=true /p:SkipIISExpressTests=false /p:SkipIISForwardsCompatibilityTests=true"
+    skipArgs: "/p:SkipIISBackwardsCompatibilityTests=true /p:SkipIISTests=true /p:SkipIISExpressTests=false /p:SkipIISForwardsCompatibilityTests=true"
```
Comment (jkotalik) [inline .azure/pipelines/ci.yml]: Do this in a separate PR.

File: src/Servers/IIS/src/AspNetCoreModuleV2/CommonLib/StringHelpers.h [hunk 2]
```diff
@@ -48,3 +51,9 @@ std::string format(const std::string& format, Args ... args)
     return result;
 }
 
+struct ignore_case_comparer
+{
+    bool operator() (const std::wstring & s1, const std::wstring & s2) const {
+        return compare_ignore_case(s1, s2) == -1;
```
Comment (jkotalik) [inline src/Servers/IIS/src/AspNetCoreModuleV2/CommonLib/StringHelpers.h]: I thought compare_ignore_case would return 0 if the values were equal. Why are we comparing against -1?

**Review by jkotalik — COMMENTED** [review]
Looks pretty good. Resolve the test failures and add *potentially* add a few environment variable tests for out of proc 😄

File: src/Servers/IIS/src/AspNetCoreModuleV2/CommonLib/StringHelpers.h (hunk 2 above)
Comment (pakrym) [reply src/Servers/IIS/src/AspNetCoreModuleV2/CommonLib/StringHelpers.h]:
```
Binary predicate that, taking two element keys as argument, returns true if the first argument goes before the second argument in the strict weak ordering it defines, and false otherwise.
```

File: src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h [hunk 3]
```diff
@@ -24,422 +26,77 @@ class ENVIRONMENT_VAR_HELPERS
...
-            delete pEnvironmentVarTable;
-            pEnvironmentVarTable = NULL;
-        }
-        return hr;
+        return environmentVariables;
     }
 
     static
-    HRESULT
+    std::map<std::wstring, std::wstring, ignore_case_comparer>
```
Comment (jkotalik) [inline src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h]: Why return the environment variable map when you already using a ref for the parameter?

**Review by jkotalik — APPROVED** [review]

File: src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h (hunk 3 above)
Comment (pakrym) [reply src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h]: It's a modified copy of the original map. It takes original as a const reference to avoid one extra copy.

Comment (jkotalik) [reply src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h]: Why not just modify the map passed in to avoid the copy?

Comment (pakrym) [reply src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h]: Because that's the original one stored in options and in OutOfProc InitEnvironmentVariablesTable might be called multiple times.

**Comment by jbeanky** [comment]
Thanks for the quick fix. Hope this gets released soon.  Missing out on our 400% throughput improvement due to this issue.

**Comment by samipedge** [comment]
Hey guys! Has this been released or will it be released soon? How would I know which version it would be in?
Thanks for all your work!

**Comment by anurse** [comment]
> Hey guys! Has this been released or will it be released soon?

This was merged as part of .NET Core 3.0, so it should be present in the latest preview builds available here: https://dotnet.microsoft.com/download/dotnet-core/3.0
